### ISSUE TITLE
Allow async in hx-trigger conditional calls.

### ISSUE THREAD

**Issue by gnat** [bigskysoftware/htmx#912]
# Use case

Hotwire recently added a similar feature. Would make available some intuitive low-code patterns in htmx. 

If `hx-trigger="submit[action()]"` supported async calls, we could "pause" the event, do something, then "resume" the event. Real life uses:

* Check an external captcha. `hx-trigger="submit[recaptcha(target)]"`
* Fetch external tokens. `hx-trigger="submit[auth(target)]"`
* Other network-dependent UX or timeline animation directly in htmx.

Suggestion to achieve this: htmx can simply detect if `hx-trigger` is a promise or vanilla function, and handle it accordingly.

# Benefits

It would totally dunk on our friends at Hotwire / Turbo to allow this. :sweat_smile: And go very well in the [Migration Guide](https://github.com/gnat/htmx/blob/hotwire_migration_guide/www/migration-guide-hotwire-turbo.md)

The Hotwire way: <https://turbo.hotwired.dev/handbook/drive#pausing-requests>

Would simplify and cut down the amount of event listener code drastically in a very htmx-way.

# Issues

Currently, you can pass an async function in, but `maybeGenerateConditional()` doesn't understand it: <https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L1001>

@1cg said:
> i've had a few comments asking about async support, but that would involve promises which are not IE compatible

**Comment by gnat** [comment]
As a side note, the Hotwire implementation of event pausing / interception: https://github.com/hotwired/turbo/pull/306

**Comment by gnat** [comment]
Someone else stumbling on this:

![image](https://user-images.githubusercontent.com/24665/225224551-74cf7cf6-d9d8-4c2b-9113-46b8469dc535.png)

**Comment by 1cg** [comment]
Is the (relatively recent) `htmx:confrm` event a satisfying solution to this: https://htmx.org/events/#htmx:confirm

Looking at the code, making all the places where filter expressions are evaluated handle async well would be a little ugly (e.g. polling and load)

**Comment by eduardvercaemer** [comment]
I was working on something that also required async event handlers (i.e. i was using `htmx:configRequest` to modify the request but it was async).

I came up with my own hack in htmx to support this logic:
```js
function triggerEvent(elt, eventName, detail, continuation /* this is new */) {
      elt = resolveTarget(elt);
      if (detail == null) {
        detail = {};
      }
      detail["elt"] = elt;
      if (continuation) { /* this is new */
        var continuationPromises = [];
        detail["continuation"] = function (eventCb) {
          var resolveContinuation = null;
          continuationPromises.push(
            new Promise(function (resolve) {
              resolveContinuation = resolve;
            })
          );
          eventCb(resolveContinuation);
        };
      }
      var event = makeEvent(eventName, detail);
      if (htmx.logger && !ignoreEventForLogging(eventName)) {
        htmx.logger(elt, eventName, detail);
      }
      if (detail.error) {
        logError(detail.error);
        triggerEvent(elt, "htmx:error", { errorInfo: detail });
      }
      var eventResult = elt.dispatchEvent(event);
      var kebabName = kebabEventName(eventName);
      if (eventResult && kebabName !== eventName) {
        var kebabedEvent = makeEvent(kebabName, event.detail);
        eventResult = eventResult && elt.dispatchEvent(kebabedEvent);
      }
      withExtensions(elt, function (extension) {
        eventResult =
          eventResult && extension.onEvent(eventName, event) !== false;
      });
      if (continuation) { /* this is new */
        return Promise.all(continuationPromises).then(function () {
          return continuation(eventResult);
        });
      }
      return eventResult;
    }
    
    // ...
    
    
      return triggerEvent(
        elt,
        "htmx:configRequest",
        requestConfig,
        afterConfiguration
      );

      function afterConfiguration(eventResult) {
        if (!eventResult) {
          maybeCall(resolve);
          endRequestLock();
          return promise;
        }
        // ...
```

and you set up a handler like this
```js

htmx.on("htmx:configRequest", (event) => {
  event.detail.continuation((resolve) => {
    // do something async
    resolve();
  });
});
```

now triggerEvent will allow listeners to register a promise and wont call the continuation untill all listeners are resolved.

**Comment by yawaramin** [comment]
Looks like this example shows how to accomplish all these use cases? https://htmx.org/examples/async-auth/

**Comment by gnat** [comment]
Correct, an "intermediate" event is the current best workaround- it's just extra steps.

The story here is DX simplification by allowing both regular and async functions to be triggered inline.

**Comment by 1cg** [comment]
wontfix - we've made these expressions sync forever now and it's been OK.  You can always use a custom event if necessary.
